### PR TITLE
Fix/multiples

### LINKED ISSUE #5840: Minecraft crashes after right clicking blackmith's hut.
### Prerequisites 

- [x] I am running the latest **alpha** version of MineColonies and Structurize for my Minecraft version.
- [x] I checked the MineColonies/Structurize [wiki](https://wiki.minecolonies.ldtteam.com/) and made sure my issue is not covered there.
- [x] I made sure that **this issue is not a duplicate of any existing issue.**

- Minecraft Version: 1.15.2
- MineColonies Version:  0.13.275-ALPHA
- Structurize Version: 0.13.64-ALPHA
- Related Mods and their Versions:

Usually when you right click on the blacksmith's hut block, it would bring you to a menu,
however, currenty whenever I try to open the hut menu, it crashes the entire game. This is currently on a minecraft multiplayer server

## Steps to reproduce the issue

1. Open your minecraft save or server
2. Find and rightclick the blacksmith's hut
3. Game will display an empty blacksmith's menu very briefly before crashing

- crashlog: https://gist.github.com/Aasterza/8e73fac79f226601365e4716f2e6d24a

### PR DESCRIPTION
**Pull request by Raycoms**
Closes #5840
Closes #
Closes #

# Changes proposed in this pull request:
Better stat inheritance
Increased pupil leveling
Warehouse resolving to closest
Remove recall all at townhall
Add recall to citizen hut
Fix team display on client side

Review please

### CODE REVIEW

File: src/main/java/com/minecolonies/coremod/colony/managers/BuildingManager.java [hunk 1]
```diff
@@ -257,6 +257,34 @@ public IBuilding getBuilding(final BlockPos buildingId)
...
+            }
+        }
+
+        wareHouses.removeIf(Objects::isNull);
+        wareHouses.sort((w1, w2) ->
+        {
+            final double dist1 = w1.getPosition().distanceSq(pos);
+            final double dist2 = w2.getPosition().distanceSq(pos);
+            return Double.compare(dist1, dist2);
+        });
```
Comment (someaddons) [inline src/main/java/com/minecolonies/coremod/colony/managers/BuildingManager.java]: if you only need the closest just keep the closest saved, no need to re-sort the entire list

File: src/main/java/com/minecolonies/coremod/colony/requestsystem/resolvers/core/AbstractWarehouseRequestResolver.java [hunk 2]
```diff
@@ -264,6 +262,14 @@ public void onAssignedRequestCancelled(@NotNull final IRequestManager manager, @
             }
         }
 
+        wareHouses.removeIf(Objects::isNull);
```
Comment (someaddons) [inline src/main/java/com/minecolonies/coremod/colony/requestsystem/resolvers/core/AbstractWarehouseRequestResolver.java]: unneeded null check, we only add notnull objects

File: src/main/java/com/minecolonies/coremod/network/messages/server/colony/building/RecallCitizenHutMessage.java [hunk 3]
```diff
@@ -0,0 +1,83 @@
...
+    }
+
+    @Override
+    protected void onExecute(@NotNull final NetworkEvent.Context ctxIn, final boolean isLogicalServer, @NotNull final IColony colony, @NotNull final IBuilding building)
+    {
+        final BlockPos location = building.getPosition();
+        final World world = colony.getWorld();
+        for (final ICitizenData citizenData : building.getAssignedCitizen())
+        {
+            Optional<AbstractEntityCitizen> optionalEntityCitizen = citizenData.getEntity();
```
Comment (someaddons) [inline src/main/java/com/minecolonies/coremod/network/messages/server/colony/building/RecallCitizenHutMessage.java:51]: could we here instead do data.setLastpos(buildingpos) -> existing entity remove(or null) -> data.updateEntityIfNecessary() ?

Then we're making sure to cleanup the references and get a valid entity in any case. E.g. when the data somehow has a no longer valid entities but still keeps that

**Comment by ravenbuilder934** [comment]
Why did you remove the recall all? It's very useful for making sure colonists aren't stuck anywhere.

**Comment by ravenbuilder934** [comment]
"Better stat inheritance" does this close https://github.com/ldtteam/minecolonies-features/issues/161?

**Comment by Raycoms** [comment]
Because recall-all only works when you have < 10 citizens and even with 10 there is a chance of them dieing.
I am adding a recall to the citizen huts instead.

**Comment by ravenbuilder934** [comment]
Dying? I've recalled ~50 citizens at the same time and none of them have died (maybe birch TH is just bigger though)

**Comment by Raycoms** [comment]
Lucky you, we lost 100 citizens these days from a recall all.

**Comment by Raycoms** [comment]
and yeah, it will close ldtteam/minecolonies-features#161

**Review by marchermans — APPROVED** [review]

**Review by Mekle001 — APPROVED** [review]

**Pull request merged**